fix: NetworkAnimator does not unsubscribe to OnClientConnectedCallback [MTT-4080]

NoelStephensUnity opened the pull request:
This resolves the issue where a NetworkBehaviour component would not be able to determine if it was a client or server when its associated NetworkObject is despawned during the shutdown sequence.

[MTT-4080](https://jira.unity3d.com/browse/MTT-4080)
Fixes #2044 

## Changelog
- Fixed issue where NetworkAnimator was not removing its subscription from OnClientConnectedCallback when despawned during the shutdown sequence. As well, IsServer and IsClient will still be valid properties to check when any NetworkObject is despawned during the shutdown sequence.

## Testing and Documentation
- Includes integration validation test.

NoelStephensUnity committed:
test

Adding the integration test that verifies NetworkAnimator properly removes its subscription to the NetworkManager.OnClientConnectedCallback when it is despawned.

NoelStephensUnity committed:
update

fixing merge issue with changelog.

NoelStephensUnity committed:
fix

MTT-4080
This resolves the issue with NetworkAnimator not unsubscribing to OnClientConnectedCallback if the NetworkObject it belongs to is despawned during the shutdown sequence.

NoelStephensUnity committed:
test

The test that validates this fix.

NoelStephensUnity committed:
Update CHANGELOG.md

updating the PR number.

JesseOlmer commented on com.unity.netcode.gameobjects/CHANGELOG.md:
```diff
@@ -5,15 +5,14 @@ All notable changes to this project will be documented in this file.
 
 ### Fixed
 
+- Fixed issue where NetworkAnimator was not removing its subscription from OnClientConnectedCallback when despawned during the shutdown sequence. As well, IsServer and IsClient will still be valid properties to check when any NetworkObject is despawned during the shutdown sequence. (#2074)
```
```suggestion
- Fixed issue where NetworkAnimator was not removing its subscription from OnClientConnectedCallback when despawned during the shutdown sequence. (#2074)
- Fixed IsServer and IsClient being set to false before object despawn during the shutdown sequence. (#2074)
```

NoelStephensUnity committed:
Update com.unity.netcode.gameobjects/CHANGELOG.md

Co-authored-by: Jesse Olmer <jesseo@unity3d.com>

NoelStephensUnity committed:
test

removing the second spawn as that was legacy code from the original fix where it removed the check within NetworkAnimator for IsServer during cleanup.
